Reorganize scheme type converter into apimachinery utils

jpbetz opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The generated apply configurations depend on a testing package for type conversion from `runtime.Object` to structured-merge-diff's `TypedValue`.  We should not have dependencies from production code to testing code.

Git blame says I added the dep.. apologies all.  I'll clean up my own mess here.

#### Special notes for your reviewer:

This will fail apidiff because the generated `NewTypeConverter` declared it's return type as a struct defined in the testing package.  This instead returns an interface that is compatible with the struct previously returned.  This is not strictly backward compatible but I expect it to be source compatible with all call sites.

This reduces the duplications `friendlyName` duplications of the `ToRESTFriendlyName` function from kube-openapi from 5 to 1.  I kept one to avoid introducing a k8s.io/apimachinery → k8s.io/kube-openapi dependency.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

jpbetz commented:
/sig api-machinery
@aojea @BenTheElder @Jefftree

aojea commented:
apidiff complains but is removing a testing dependency from production binaries

```
## ./staging/src/k8s.io/code-generator/examples
Incompatible changes:
- ./single/applyconfiguration.NewTypeConverter: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) *k8s.io/client-go/testing.TypeConverter to func(*k8s.io/apimachinery/pkg/runtime.Scheme) k8s.io/apimachinery/pkg/util/managedfields.TypeConverter
- ./HyphenGroup/applyconfiguration.NewTypeConverter: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) *k8s.io/client-go/testing.TypeConverter to func(*k8s.io/apimachinery/pkg/runtime.Scheme) k8s.io/apimachinery/pkg/util/managedfields.TypeConverter
- ./crd/applyconfiguration.NewTypeConverter: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) *k8s.io/client-go/testing.TypeConverter to func(*k8s.io/apimachinery/pkg/runtime.Scheme) k8s.io/apimachinery/pkg/util/managedfields.TypeConverter
- ./MixedCase/applyconfiguration.NewTypeConverter: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) *k8s.io/client-go/testing.TypeConverter to func(*k8s.io/apimachinery/pkg/runtime.Scheme) k8s.io/apimachinery/pkg/util/managedfields.TypeConverter
## ./staging/src/k8s.io/client-go
Incompatible changes:
- ./applyconfigurations.NewTypeConverter: changed from func(*k8s.io/apimachinery/pkg/runtime.Scheme) *k8s.io/client-go/testing.TypeConverter to func(*k8s.io/apimachinery/pkg/runtime.Scheme) k8s.io/apimachinery/pkg/util/managedfields.TypeConverter
- ./testing.TypeConverter: removed
- ./testing.TypeResolver: removed

```

aojea commented on pkg/generated/openapi/cmd/models-schema/main.go:
```diff
@@ -19,12 +19,11 @@ package main
 	"k8s.io/kube-openapi/pkg/common"
 	"k8s.io/kube-openapi/pkg/validation/spec"
 	"k8s.io/kubernetes/pkg/generated/openapi"
+	"os"
```
nit , usually the stdlib imports go first and separated

jpbetz replied:
gah, my IDE's import settings got reset. Fixing..

aojea commented on staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/openapi.go:
```diff
@@ -19,6 +19,7 @@ package generators
 import (
 	"encoding/json"
 	"fmt"
+	"k8s.io/kube-openapi/pkg/util"
```
heh, your IDE was mixing the imports

aojea commented on pkg/generated/openapi/cmd/models-schema/main.go:
```diff
@@ -19,12 +19,11 @@ package main
-	"os"
-	"strings"
-
+	"k8s.io/apimachinery/pkg/runtime"
```
was this on purpose before to avoid this dependency?

cc: @liggitt

liggitt replied:
not sure what the original intent was, but we do already have this indirectly via k8s.io/kubernetes/pkg/generated/openapi → k8s.io/apimachinery/pkg/apis/meta/v1 → k8s.io/apimachinery/pkg/runtime

jpbetz replied:
Previously there were 4 private copies of the [ToRESTFriendlyName](https://github.com/kubernetes/kube-openapi/blob/c8a335a9a2ffc5aff16dfef74896a1ee34eb235d/pkg/util/util.go#L58) function, all named `friendyName`.  This reduces that down to one in k8s.io/apimachinery/pkg/runtime.

I kept the last one to avoid adding a dependency from k8s.io/apimachinery → k8s.io/kube-openapi which results dependencies like k8s.io/api → k8s.io/apimachinery → k8s.io/kube-openapi. I didn't want to add that.

Here's the last remaining copy:

https://github.com/kubernetes/kubernetes/pull/131616/files#diff-7ac25bcc6f0498d4ade98f19cfd251d2cb4cdfbe3c21a72eb7ba8eb0b0f4fdd0R771-R773

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -743,3 +743,42 @@ func (s *Scheme) Name() string {
+//
+// Note that this is a copy of ToRESTFriendlyName from k8s.io/kube-openapi/pkg/util. It is duplicated here to avoid
+// a dependency on kube-openapi.
+func ToRESTFriendlyName(name string) string {
```
We currently have 5 copies of these, so collapsing them so they don't drift sounds like a good idea

I would name the exported method more precisely, something like `ToOpenAPIDefinitionName`, and document the input is expected to be `PkgPath + "." TypeName

jpbetz replied:
I was able to make this one private.  Many of the other places where we needed this function we already have a dep on kube-openapi so I've updated to call the kube-openapi `ToRESTFriendlyName` directly in those cases and only keep one private copy for runtime.

aaron-prindle commented:
/triage accepted

jpbetz commented:
@liggitt @aojea I think this is ready now.

BenTheElder commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -743,3 +743,43 @@ func (s *Scheme) Name() string {
+//	Input:  csi.storage.k8s.io/v1alpha1.CSINodeInfo
+//	Output: io.k8s.storage.csi.v1alpha1.CSINodeInfo
+//
+// Note that this is a copy of ToRESTFriendlyName from k8s.io/kube-openapi/pkg/util. It is duplicated here to avoid
```
what skew risk do we have? should we put an "if this then that" style comment in kube-openapi, or try to move this to a smaller additional module?

jpbetz replied:
It feels relatively low.  This function has been stable for >>years.  Long enough that we accumulated 5 copies of it in k/k.  It's coded against API guarantees so it's really shouldn't change.

I'd still like to somehow get this declaration into a common lib, but I don't think risk of change is practical concern.

BenTheElder commented:
/priority important-longterm

aojea commented:
/lgtm

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -743,3 +743,43 @@ func (s *Scheme) Name() string {
+	if err != nil {
+		return "", err
+	}
+	rtype := reflect.TypeOf(example).Elem()
```
If we're making this more official / available, do we need to protect against ending up with an unstructured.Unstructured type here?

I think scheme makes it *possible* to do things like `unstructuredScheme.AddKnownTypeWithName(gvk, &unstructured.Unstructured{})`

jpbetz replied:
Good catch.

I'd supported unstructured and documented that unstructured is assumed to follow custom resource convensions

As a sanity check, I verified that in our OpenAPI something like `/apis/stable.example.com/v1/crontabs` maps to `#/components/schemas/com.example.stable.v1.CronTab`.

Also added examples to the godoc and a couple test cases.

liggitt commented on staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/types.go:
```diff
@@ -19,18 +19,19 @@ package generators
-	runtimeScheme        = types.Ref("k8s.io/apimachinery/pkg/runtime", "Scheme")
-	smdNewParser         = types.Ref("sigs.k8s.io/structured-merge-diff/v4/typed", "NewParser")
-	smdParser            = types.Ref("sigs.k8s.io/structured-merge-diff/v4/typed", "Parser")
-	testingTypeConverter = types.Ref("k8s.io/client-go/testing", "TypeConverter")
```
is there anything we can add to make sure this doesn't creep back into non-test binaries? maybe in hack/verify-testing-import.sh

jpbetz replied:
Opened https://github.com/kubernetes/kubernetes/issues/131646 to track.  I'll try to get it assigned.

liggitt commented:
one question on unstructured handling and one on a test to prevent regressing, lgtm overall (probably a good idea to squash the non-generated commits together)

jpbetz commented:
> one question on unstructured handling and one on a test to prevent regressing, lgtm overall (probably a good idea to squash the non-generated commits together)

CRDs can be fully supported. I'm adding the support and the test cases now.  Then I'll squash.  I'll open an issue to track the need for better linting.

jpbetz committed:
Reorganize scheme type converter into apimachinery utils

This removes a dependency from generated applyconfigurations to a testing
package. To do this, the type converter in the testing package has been
moved out to the apimachinery package and the utilities the converter
depend on have been reorganized.

jpbetz committed:
generate code

jpbetz commented:
/retest

jpbetz commented:
@liggitt this is ready

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -743,3 +743,59 @@ func (s *Scheme) Name() string {
+		return "", err
+	}
+	if _, ok := example.(Unstructured); ok {
+		return reverseParts(groupVersionKind.Group) + "." + groupVersionKind.Version + "." + groupVersionKind.Kind, nil
```
what happens here if `group` is empty? is it possible for `version` or `kind` to be empty and `New(groupVersionKind)` to have succeeded above?

jpbetz replied:
Added tests and added code to guard against empty GVK parts.  `New` panics specifically on empty version, so the guards really help in that case.

liggitt commented:
/lgtm
/approve
